tests/unit/s2n_mutual_auth_test.c

```diff
@@ -0,0 +1,349 @@
+{
+    struct s2n_stuffer *out;
+
+    if (buf == NULL) {
+        return 0;
+    }
+
+    out = (struct s2n_stuffer *) io_context;
+    if (out == NULL) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    if (s2n_stuffer_write_bytes(out, buf, len) < 0) {
+        errno = EAGAIN;
+        return -1;
+    }
+
+    return len;
+}
+
+extern message_type_t s2n_conn_get_current_message_type(struct s2n_connection *conn);
+
+static const int MAX_TRIES = 100;
+
+int main(int argc, char **argv)
+{
+    struct s2n_config *config;
+    const struct s2n_cipher_preferences *default_cipher_preferences;
+
+
+    BEGIN_TEST();
+    EXPECT_SUCCESS(setenv("S2N_ENABLE_CLIENT_MODE", "1", 0));
+
+    /*
+     * Test Mutual Auth using **s2n_connection_set_client_auth_type**
+     */
+
+    EXPECT_NOT_NULL(config = s2n_config_new());
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key(config, certificate_chain, private_key));
+    EXPECT_SUCCESS(s2n_config_add_dhparams(config, dhparams));
+    EXPECT_NOT_NULL(default_cipher_preferences = config->cipher_preferences);
+
+    EXPECT_SUCCESS(s2n_config_set_verify_cert_chain_cb(config, &accept_all_rsa_certs, NULL));
+    EXPECT_SUCCESS(s2n_config_set_verify_cert_chain_cb(config, &accept_all_rsa_certs, NULL));
+
+    /* Verify that a handshake succeeds for every cipher in the default list. */
+    for (int cipher_idx = 0; cipher_idx < default_cipher_preferences->count; cipher_idx++) {
+        struct s2n_cipher_preferences server_cipher_preferences;
+        struct s2n_connection *client_conn;
+        struct s2n_connection *server_conn;
+        s2n_blocked_status client_blocked;
+        s2n_blocked_status server_blocked;
+        struct s2n_stuffer client_to_server;
+        struct s2n_stuffer server_to_client;
+
+        /* Craft a cipher preference with a cipher_idx cipher
+           NOTE: Its safe to use memcpy as the address of server_cipher_preferences
+           will never be NULL */
+        memcpy(&server_cipher_preferences, default_cipher_preferences, sizeof(server_cipher_preferences));
+        server_cipher_preferences.count = 1;
+        struct s2n_cipher_suite *cur_cipher = default_cipher_preferences->suites[cipher_idx];
+
+        if (!cur_cipher->available) {
+            /* Skip Ciphers that aren't supported with the linked libcrypto */
+            continue;
+        }
+
+        server_cipher_preferences.suites = &cur_cipher;
+        config->cipher_preferences = &server_cipher_preferences;
+
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        EXPECT_SUCCESS(s2n_connection_set_client_auth_type(server_conn, S2N_CERT_AUTH_REQUIRED));
+        EXPECT_SUCCESS(s2n_connection_set_client_auth_type(client_conn, S2N_CERT_AUTH_REQUIRED));
```

Comment:
This line here is an example of my previous comments about the API. It seems confusing to me to be setting the client auth type on the client. I feel like the server is the only side that needs to be configured, and the client should just be reactive. For now, it doesn't seem like a big deal. But, when we start adding more complicated options for sending the CertificateRequest (e.g. set on config and override on connection, change the auth requirements in the post-ClientHello callback, etc) it could be difficult to ensure that both client and server match in configuration. Does the client need to validate that the CertificateRequest was sent as expected? Or do you anticipate there being a case where the client will choose not to send the certificate even if the CertificateRequest is sent? If so, then maybe this makes sense.